backends/tc/run-tc-test.py

```diff
@@ -25,225 +26,155 @@
+def run_test(options, argv):
+    """Define the test environment and compile the p4 target
+    Optional: Run the generated model"""
     assert isinstance(options, Options)
 
-    tmpdir = tempfile.mkdtemp(dir=Path(".").absolute())
-    basename = os.path.basename(options.p4filename)
-    base, ext = os.path.splitext(basename)
-    dirname = os.path.dirname(options.p4filename)
-    expected_dirname = dirname + "_outputs"  # expected outputs are here
-
-    if options.verbose:
-        print("Writing temporary files into ", tmpdir)
-    outputfolder = tmpdir + "/"
-    stderr = tmpdir + "/" + basename + "-stderr"
-
-    if not os.path.isfile(options.p4filename):
-        raise Exception("No such file " + options.p4filename)
-    args = ["./p4c-pna-p4tc", "-o", outputfolder]
-    args.extend(argv)
-    print("input: ", options, args, timeout, stderr)
-    result = run_timeout(options, args, timeout, stderr)
-
-    if result != SUCCESS:
-        print("Error compiling")
-        print(" ".join(open(stderr).readlines()))
-        # If the compiler crashed fail the test
-        if 'Compiler Bug' in open(stderr).readlines():
-            return FAILURE
-
-    expected_error = isError(options.p4filename)
-    if expected_error:
-        # invert result
-        if result == SUCCESS:
-            result = FAILURE
-        else:
-            result = SUCCESS
-
-    if result == SUCCESS:
-        result = check_generated_files(options, tmpdir, expected_dirname)
-
-    if options.cleanupTmp:
-        if options.verbose:
-            print("Removing", tmpdir)
-        shutil.rmtree(tmpdir)
-    return result
+    template, _ = os.path.splitext(options.p4filename)
+    tmpdir = tempfile.mkdtemp(dir=Path(FILE_DIR.joinpath("runtime")).absolute())
+    outputfolder = tmpdir
 
+    tc = TCInfra(outputfolder, options, template)
 
-def isdir(path):
-    try:
-        return stat.S_ISDIR(os.stat(path).st_mode)
-    except OSError:
-        return False
+    return run_model(tc, options.testfile)
 
 
 ######################### main
 
 
 def main(argv):
+    args, argv = PARSER.parse_known_args()
     options = Options()
-    print("P4TC testing")
-    options.binary = argv[0]
-    if len(argv) <= 2:
-        usage(options)
-        sys.exit(FAILURE)
-
-    options.compilerSrcdir = argv[1]
-    argv = argv[2:]
-    if not os.path.isdir(options.compilerSrcdir):
-        print(options.compilerSrcdir + " is not a folder", file=sys.stderr)
-        usage(options)
-        sys.exit(FAILURE)
-
-    while argv[0][0] == '-':
-        if argv[0] == "-b":
-            options.cleanupTmp = False
-        elif argv[0] == "-v":
-            options.verbose = True
-        elif argv[0] == "-f":
-            options.replace = True
-        else:
-            print("Unknown option ", argv[0], file=sys.stderr)
-            usage(options)
-        argv = argv[1:]
+    options.compiler_src_dir = testutils.check_if_dir(args.compiler_src_dir).as_posix()
```

Comment:
I would consider using Python's pathlib utilities instead of using raw strings. Makes it much easier to deal with files like this.

Reply:
Thanks, will take a look at it

Reply:
Fixed